### ISSUE TITLE
/rotation is slow because of read_files and prepare_card

### ISSUE THREAD

**Issue by bakert** [PennyDreadfulMTG/Penny-Dreadful-Tools#5471]

**Comment by tsutton** [comment]
`read_files` is mostly handled by #6688 . As far as `prepare_card`, I feel like the crux of the problem is that the function is super overloading, preparing cards for what any card might need on any page. Maybe it should be split up with a view being able to configure which parts of cards need preparing? As an extremely barebones proof of concept, if I comment out various parts of `prepare_card` I can get it down by quite a bit for `/rotation`, but there might be some overhead if there's various checks against which components are needed.

The other idea would be to try and batch the card preparation, i.e. right now we call `url_for` for each card, whereas we could call it once for all cards to get a format string and plug the cardname into that.

The other thing is, it takes a long time for pystache to render all the cards. Locally, on a hard refresh, it is the dominant component by far. This may not be a realistic representation of prod, though, f it would be helped by the `CachedRenderer` magic in https://github.com/PennyDreadfulMTG/Penny-Dreadful-Tools/blob/master/shared_web/template.py if I weren't doing a hard refresh. Apparently `pystache` is basically abandoned, and https://github.com/noahmorrison/chevron is supposed to be faster, but chevron is much more basic, and it might be hard to migrate over.

**Comment by tsutton** [comment]
For reference, here's what I'm looking at: three runs of `/rotation/` with 11012 cards each in 50 identical `Run_*.txt` files. Redis fully primed, code set to after #6688 .
```
** 1 - 4.98s
PATH: '/rotation/'
         6459734 function calls (5150999 primitive calls) in 4.976 seconds

   Ordered by: internal time, call count
   List reduced from 786 to 30 due to restriction <30>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   418456    0.403    0.000    0.403    0.000 /usr/lib/python3.7/site-packages/munch/__init__.py:109(__setattr__)
        3    0.374    0.125    0.374    0.125 /usr/lib/python3.7/json/decoder.py:343(raw_decode)
    11012    0.197    0.000    1.343    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/view.py:221(prepare_card)
  50760/1    0.144    0.000    3.646    3.646 /usr/lib/python3.7/site-packages/pystache/parsed.py:37(render)
 332121/4    0.127    0.000    3.645    0.911 /usr/lib/python3.7/site-packages/pystache/parsed.py:43(get_unicode)
    11035    0.127    0.000    0.127    0.000 {built-in method posix.stat}
146150/329    0.124    0.000    3.539    0.011 /usr/lib/python3.7/site-packages/pystache/context.py:203(get)
    22125    0.121    0.000    0.483    0.000 /usr/lib/python3.7/site-packages/werkzeug/routing.py:1909(build)
146149/328    0.119    0.000    3.537    0.011 /usr/lib/python3.7/site-packages/pystache/context.py:304(_get_simple)
 33170/11    0.118    0.000    2.176    0.198 /usr/lib/python3.7/site-packages/pystache/parser.py:191(render)
184517/350    0.117    0.000    3.535    0.010 /usr/lib/python3.7/site-packages/pystache/context.py:37(_get_value)
    22125    0.114    0.000    0.841    0.000 /usr/lib/python3.7/site-packages/flask/helpers.py:226(url_for)
11154/11148    0.095    0.000    0.154    0.000 {method 'sub' of 're.Pattern' objects}
   539666    0.090    0.000    0.097    0.000 {built-in method builtins.isinstance}
90955/204    0.085    0.000    3.521    0.017 /usr/lib/python3.7/site-packages/pystache/renderengine.py:100(fetch_string)
    90983    0.080    0.000    0.291    0.000 /usr/lib/python3.7/site-packages/pystache/renderer.py:191(_escape_to_unicode)
    90983    0.074    0.000    0.142    0.000 /usr/lib/python3.7/html/__init__.py:12(escape)
   264442    0.074    0.000    0.294    0.000 {built-in method builtins.setattr}
    90952    0.072    0.000    0.812    0.000 /usr/lib/python3.7/site-packages/pystache/parser.py:120(render)
161512/161506    0.071    0.000    0.072    0.000 {method 'join' of 'str' objects}
   464748    0.070    0.000    0.070    0.000 {method 'replace' of 'str' objects}
        1    0.065    0.065    1.259    1.259 /home/taylors/projects/Penny-Dreadful-Tools/decksite/views/rotation.py:13(__init__)
    66630    0.060    0.000    0.113    0.000 /usr/lib/python3.7/site-packages/werkzeug/local.py:163(top)
    11012    0.057    0.000    0.362    0.000 /home/taylors/projects/Penny-Dreadful-Tools/magic/models/card.py:8(__init__)
146150/329    0.057    0.000    3.539    0.011 /usr/lib/python3.7/site-packages/pystache/renderengine.py:14(context_get)
146150/329    0.052    0.000    3.539    0.011 /usr/lib/python3.7/site-packages/pystache/renderer.py:315(resolve_context)
44250/22125    0.051    0.000    0.257    0.000 /usr/lib/python3.7/site-packages/werkzeug/routing.py:1887(_partial_build)
        1    0.044    0.044    4.965    4.965 /home/taylors/projects/Penny-Dreadful-Tools/decksite/cache.py:31(decorated_function)
    66634    0.042    0.000    0.053    0.000 /usr/lib/python3.7/site-packages/werkzeug/local.py:70(__getattr__)
   158985    0.042    0.000    0.042    0.000 /home/taylors/projects/Penny-Dreadful-Tools/shared/container.py:9(__getattr__)


** 2 - 5.20
PATH: '/rotation/'
         6459723 function calls (5150988 primitive calls) in 5.203 seconds

   Ordered by: internal time, call count
   List reduced from 786 to 30 due to restriction <30>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   418456    0.433    0.000    0.433    0.000 /usr/lib/python3.7/site-packages/munch/__init__.py:109(__setattr__)
        3    0.324    0.108    0.324    0.108 /usr/lib/python3.7/json/decoder.py:343(raw_decode)
    11012    0.197    0.000    1.312    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/view.py:221(prepare_card)
  50760/1    0.170    0.000    3.870    3.870 /usr/lib/python3.7/site-packages/pystache/parsed.py:37(render)
 332121/4    0.148    0.000    3.867    0.967 /usr/lib/python3.7/site-packages/pystache/parsed.py:43(get_unicode)
    11035    0.141    0.000    0.141    0.000 {built-in method posix.stat}
146149/328    0.141    0.000    3.815    0.012 /usr/lib/python3.7/site-packages/pystache/context.py:304(_get_simple)
146150/329    0.138    0.000    3.815    0.012 /usr/lib/python3.7/site-packages/pystache/context.py:203(get)
 33170/11    0.136    0.000    2.481    0.226 /usr/lib/python3.7/site-packages/pystache/parser.py:191(render)
184517/350    0.134    0.000    3.814    0.011 /usr/lib/python3.7/site-packages/pystache/context.py:37(_get_value)
    22125    0.117    0.000    0.465    0.000 /usr/lib/python3.7/site-packages/werkzeug/routing.py:1909(build)
    22125    0.111    0.000    0.816    0.000 /usr/lib/python3.7/site-packages/flask/helpers.py:226(url_for)
11154/11148    0.106    0.000    0.174    0.000 {method 'sub' of 're.Pattern' objects}
   539665    0.095    0.000    0.101    0.000 {built-in method builtins.isinstance}
90955/204    0.094    0.000    3.807    0.019 /usr/lib/python3.7/site-packages/pystache/renderengine.py:100(fetch_string)
    90983    0.090    0.000    0.326    0.000 /usr/lib/python3.7/site-packages/pystache/renderer.py:191(_escape_to_unicode)
        1    0.086    0.086    1.293    1.293 /home/taylors/projects/Penny-Dreadful-Tools/decksite/views/rotation.py:13(__init__)
    90952    0.084    0.000    0.908    0.000 /usr/lib/python3.7/site-packages/pystache/parser.py:120(render)
    90983    0.083    0.000    0.159    0.000 /usr/lib/python3.7/html/__init__.py:12(escape)
161512/161506    0.081    0.000    0.081    0.000 {method 'join' of 'str' objects}
   464748    0.077    0.000    0.077    0.000 {method 'replace' of 'str' objects}
   264442    0.076    0.000    0.298    0.000 {built-in method builtins.setattr}
146150/329    0.065    0.000    3.815    0.012 /usr/lib/python3.7/site-packages/pystache/renderengine.py:14(context_get)
    11012    0.059    0.000    0.371    0.000 /home/taylors/projects/Penny-Dreadful-Tools/magic/models/card.py:8(__init__)
    66630    0.059    0.000    0.113    0.000 /usr/lib/python3.7/site-packages/werkzeug/local.py:163(top)
146150/329    0.059    0.000    3.816    0.012 /usr/lib/python3.7/site-packages/pystache/renderer.py:315(resolve_context)
    44048    0.053    0.000    0.072    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/views/rotation.py:56(redact)
   158985    0.051    0.000    0.051    0.000 /home/taylors/projects/Penny-Dreadful-Tools/shared/container.py:9(__getattr__)
44250/22125    0.051    0.000    0.249    0.000 /usr/lib/python3.7/site-packages/werkzeug/routing.py:1887(_partial_build)
    33170    0.048    0.000    0.263    0.000 /usr/lib/python3.7/site-packages/pystache/renderengine.py:116(fetch_section_data)


** 3 - 4.72
PATH: '/rotation/'
         6459723 function calls (5150988 primitive calls) in 4.716 seconds

   Ordered by: internal time, call count
   List reduced from 786 to 30 due to restriction <30>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   418456    0.379    0.000    0.379    0.000 /usr/lib/python3.7/site-packages/munch/__init__.py:109(__setattr__)
    11012    0.177    0.000    1.348    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/view.py:221(prepare_card)
    11066    0.176    0.000    0.176    0.000 {built-in method builtins.sorted}
    11035    0.161    0.000    0.161    0.000 {built-in method posix.stat}
        3    0.161    0.054    0.161    0.054 /usr/lib/python3.7/json/decoder.py:343(raw_decode)
  50760/1    0.145    0.000    3.629    3.629 /usr/lib/python3.7/site-packages/pystache/parsed.py:37(render)
 332121/4    0.128    0.000    3.627    0.907 /usr/lib/python3.7/site-packages/pystache/parsed.py:43(get_unicode)
146150/329    0.120    0.000    3.587    0.011 /usr/lib/python3.7/site-packages/pystache/context.py:203(get)
146149/328    0.119    0.000    3.586    0.011 /usr/lib/python3.7/site-packages/pystache/context.py:304(_get_simple)
184517/350    0.118    0.000    3.586    0.010 /usr/lib/python3.7/site-packages/pystache/context.py:37(_get_value)
 33170/11    0.117    0.000    2.223    0.202 /usr/lib/python3.7/site-packages/pystache/parser.py:191(render)
    22125    0.108    0.000    0.426    0.000 /usr/lib/python3.7/site-packages/werkzeug/routing.py:1909(build)
    22125    0.102    0.000    0.748    0.000 /usr/lib/python3.7/site-packages/flask/helpers.py:226(url_for)
11154/11148    0.099    0.000    0.159    0.000 {method 'sub' of 're.Pattern' objects}
   539665    0.089    0.000    0.096    0.000 {built-in method builtins.isinstance}
90955/204    0.085    0.000    3.583    0.018 /usr/lib/python3.7/site-packages/pystache/renderengine.py:100(fetch_string)
    90983    0.077    0.000    0.289    0.000 /usr/lib/python3.7/site-packages/pystache/renderer.py:191(_escape_to_unicode)
   264442    0.077    0.000    0.301    0.000 {built-in method builtins.setattr}
    90983    0.075    0.000    0.143    0.000 /usr/lib/python3.7/html/__init__.py:12(escape)
    90952    0.073    0.000    0.805    0.000 /usr/lib/python3.7/site-packages/pystache/parser.py:120(render)
   464748    0.070    0.000    0.070    0.000 {method 'replace' of 'str' objects}
161512/161506    0.063    0.000    0.064    0.000 {method 'join' of 'str' objects}
    11012    0.061    0.000    0.374    0.000 /home/taylors/projects/Penny-Dreadful-Tools/magic/models/card.py:8(__init__)
        1    0.061    0.061    1.046    1.046 /home/taylors/projects/Penny-Dreadful-Tools/decksite/views/rotation.py:13(__init__)
146150/329    0.058    0.000    3.587    0.011 /usr/lib/python3.7/site-packages/pystache/renderengine.py:14(context_get)
    66630    0.053    0.000    0.101    0.000 /usr/lib/python3.7/site-packages/werkzeug/local.py:163(top)
146150/329    0.053    0.000    3.587    0.011 /usr/lib/python3.7/site-packages/pystache/renderer.py:315(resolve_context)
44250/22125    0.046    0.000    0.226    0.000 /usr/lib/python3.7/site-packages/werkzeug/routing.py:1887(_partial_build)
    33170    0.041    0.000    0.224    0.000 /usr/lib/python3.7/site-packages/pystache/renderengine.py:116(fetch_section_data)
   130510    0.039    0.000    0.039    0.000 {method 'get' of 'dict' objects}

```

**Comment by tsutton** [comment]
Here's a breakdown of `prepare_card` as called by `/rotation/` (so that actually most common cards is not doing anything). I split it out into a few different functions locally to test.
```
    14536    0.065    0.000    1.470    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/view.py:250(prepare_card)
    14536    0.058    0.000    0.192    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/view.py:223(prepare_card_most_common_cards)
    14536    0.067    0.000    0.179    0.000 /home/taylors/projects/Penny-Dreadful-Tools/decksite/view.py:237(prepare_card_legal_formats)
```
...and our champion:
```
    29173    0.126    0.000    0.895    0.000 /usr/lib/python3.7/site-packages/flask/helpers.py:226(url_for)
```
Awkward. I can save a lot of time by adding code like this to the main class in `view.py`:
```python
    def url_for_card(self, c: Card) -> str:
        if getattr(self, '_card_url_template', None) is None:
            if self.tournament_only:
                self._card_url_template = url_for('.card_tournament', name="{cardname}")
            else:
                self._card_url_template = url_for('.card', name="{cardname}")
        return self._card_url_template.format(cardname=c.name)
```
and replacing the appropriate calls to `url_for` with these, and same for `url_for("image",...)`. This makes me mildly uncomfortable, though.

The other option is to just not send all 10000-15000 cards all at once, maybe send a few at the beginning and load the rest asynchronously.

**Comment by bakert** [comment]
We could give the rotation page the React treatment that long lists of decklists currently have. I feel like mostly you want to scroll to the "action" aka cards that just made it in, just got confirmed out, are nearly in, or are nearly out. Perhaps we should redesign the whole screen to show mostly this information and allow you to browse the confirmed, had-hits-but-impossible and far-off-but-still-possible cards by clicking.

**Comment by bakert** [comment]
/rotation now has the react treatment. It ain't speedy but it's a lot better.
